File: minimal-blink-c-with-rust/blink.c
/**
 * Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include <stdbool.h>
#include <stdint.h>

#include "rust-lib/rust-lib.h"

// --------------------------------------------------------------------------------------

typedef volatile uint32_t io_rw_32;
typedef const volatile uint32_t io_ro_32;
typedef volatile uint32_t io_wo_32;

#define IO_BANK0_BASE 0x40014000u
#define PADS_BANK0_BASE 0x4001c000u
#define SIO_BASE 0xd0000000u
#define TIMER_BASE 0x40054000u

#define PADS_BANK0_GPIO0_IE_BITS 0x00000040u
#define PADS_BANK0_GPIO0_OD_BITS 0x00000080u
#define IO_BANK0_GPIO0_CTRL_FUNCSEL_LSB 0u

// --------------------------------------------------------------------------------------

#define IO_BANK0_GPIO0_STATUS_OFFSET 0x00000000

typedef struct {
    io_rw_32 status;
    io_rw_32 ctrl;
} iobank0_hw_io_t;

static iobank0_hw_io_t *const iobank0_io = (iobank0_hw_io_t*)(IO_BANK0_BASE + IO_BANK0_GPIO0_STATUS_OFFSET);

#define PADS_BANK0_GPIO0_OFFSET 0x00000004

static io_rw_32 *const padsbank0_io = (io_rw_32*)(PADS_BANK0_BASE + PADS_BANK0_GPIO0_OFFSET);

#define SIO_GPIO_OUT_SET_OFFSET 0x00000014
#define SIO_GPIO_OUT_CLR_OFFSET 0x00000018
#define SIO_GPIO_OE_SET_OFFSET 0x00000024

static io_rw_32 *const sio_gpio_out_set = (io_rw_32*)(SIO_BASE + SIO_GPIO_OUT_SET_OFFSET);
static io_rw_32 *const sio_gpio_out_clr = (io_rw_32*)(SIO_BASE + SIO_GPIO_OUT_CLR_OFFSET);
static io_rw_32 *const sio_gpio_oe_set = (io_rw_32*)(SIO_BASE + SIO_GPIO_OE_SET_OFFSET);

#define TIMER_TIMERAWL_OFFSET 0x00000028

static io_ro_32 *const timer_timerawl = (io_ro_32*)(TIMER_BASE + TIMER_TIMERAWL_OFFSET);

// --------------------------------------------------------------------------------------

#define GPIO_FUNC_SIO 5

// --------------------------------------------------------------------------------------

#define MY_PICO_DEFAULT_LED_PIN 25

void gpio_init_as_sio_output(uint32_t gpio) {
    // Set input enable on, output disable off
    padsbank0_io[gpio] = (padsbank0_io[gpio] & (PADS_BANK0_GPIO0_IE_BITS | PADS_BANK0_GPIO0_OD_BITS)) | PADS_BANK0_GPIO0_IE_BITS;

    // Zero all fields apart from fsel; we want this IO to do what the peripheral tells it.
    // This doesn't affect e.g. pullup/pulldown, as these are in pad controls.
    iobank0_io[gpio].ctrl = GPIO_FUNC_SIO << IO_BANK0_GPIO0_CTRL_FUNCSEL_LSB;

    // Setup GPIO as output
    *sio_gpio_out_clr = 1ul << gpio;
    *sio_gpio_oe_set = 1ul << gpio;
}

void my_gpio_put(uint32_t gpio, bool value) {
    uint32_t mask = 1ul << gpio;
    if (value)
        *sio_gpio_out_set = mask;
    else
        *sio_gpio_out_clr = mask;
}

void my_sleep_ms(uint32_t delay_ms) {
    uint32_t base = *timer_timerawl;
    while ((*timer_timerawl - base) < (delay_ms * 1000)) {
    }
}

int main() {
    const uint32_t DELAY_MS = get_delay_value(); // Call Rust code

    const uint32_t LED_PIN = MY_PICO_DEFAULT_LED_PIN;
    gpio_init_as_sio_output(LED_PIN);
    while (true) {
        my_gpio_put(LED_PIN, 1);
        my_sleep_ms(DELAY_MS);
        my_gpio_put(LED_PIN, 0);
        my_sleep_ms(DELAY_MS);
    }
}

File: minimal-blink-rust-bare/src/blink.rs
const NUMBER_OF_GPIOS: usize = 30;

const RESETS_BASE: usize = 0x4000c000;
const IO_BANK0_BASE: usize = 0x40014000;
const PADS_BANK0_BASE: usize = 0x4001c000;
const SIO_BASE: usize = 0xd0000000;
const TIMER_BASE: usize = 0x40054000;

const RESETS_RESET_OFFSET: usize = 0x00000000;
const RESETS_RESET_DONE_OFFSET: usize = 0x00000008;
const IO_BANK0_GPIO0_STATUS_OFFSET: usize = 0x00000000;
const PADS_BANK0_GPIO0_OFFSET: usize = 0x00000004;
const SIO_GPIO_OUT_SET_OFFSET: usize = 0x00000014;
const SIO_GPIO_OUT_CLR_OFFSET: usize = 0x00000018;
const SIO_GPIO_OE_SET_OFFSET: usize = 0x00000024;
const TIMER_TIMERAWL_OFFSET: usize = 0x00000028;

//const RESETS_RESET_USBCTRL_BITS: u32 = 0x01000000;
//const RESETS_RESET_UART1_BITS: u32 = 0x00800000;
//const RESETS_RESET_UART0_BITS: u32 = 0x00400000;
const RESETS_RESET_TIMER_BITS: u32 = 0x00200000;
//const RESETS_RESET_TBMAN_BITS: u32 = 0x00100000;
//const RESETS_RESET_SYSINFO_BITS: u32 = 0x00080000;
//const RESETS_RESET_SYSCFG_BITS: u32 = 0x00040000;
//const RESETS_RESET_SPI1_BITS: u32 = 0x00020000;
//const RESETS_RESET_SPI0_BITS: u32 = 0x00010000;
//const RESETS_RESET_RTC_BITS: u32 = 0x00008000;
//const RESETS_RESET_PWM_BITS: u32 = 0x00004000;
//const RESETS_RESET_PLL_USB_BITS: u32 = 0x00002000;
//const RESETS_RESET_PLL_SYS_BITS: u32 = 0x00001000;
//const RESETS_RESET_PIO1_BITS: u32 = 0x00000800;
const RESETS_RESET_PIO0_BITS: u32 = 0x00000400;
//const RESETS_RESET_PADS_QSPI_BITS: u32 = 0x00000200;
const RESETS_RESET_PADS_BANK0_BITS: u32 = 0x00000100;
//const RESETS_RESET_JTAG_BITS: u32 = 0x00000080;
//const RESETS_RESET_IO_QSPI_BITS: u32 = 0x00000040;
const RESETS_RESET_IO_BANK0_BITS: u32 = 0x00000020;
//const RESETS_RESET_I2C1_BITS: u32 = 0x00000010;
//const RESETS_RESET_I2C0_BITS: u32 = 0x00000008;
//const RESETS_RESET_DMA_BITS: u32 = 0x00000004;
//const RESETS_RESET_BUSCTRL_BITS: u32 = 0x00000002;
//const RESETS_RESET_ADC_BITS: u32 = 0x00000001;

const PADS_BANK0_GPIO0_IE_BITS: u32 = 0x00000040;
const PADS_BANK0_GPIO0_OD_BITS: u32 = 0x00000080;
const IO_BANK0_GPIO0_CTRL_FUNCSEL_LSB: u32 = 0;

const GPIO_FUNC_SIO: u32 = 5;

const PICO_DEFAULT_LED_PIN: usize = 25;

#[repr(C)]
struct IoBank0HwIo {
    status: u32,
    ctrl: u32,
}

struct Peripherals {
    resets_reset: *mut u32,
    resets_reset_done: *const u32,
    iobank0_io: *mut [IoBank0HwIo; NUMBER_OF_GPIOS],
    padsbank0_io: *mut [u32; NUMBER_OF_GPIOS],
    sio_gpio_out_set: *mut u32,
    sio_gpio_out_clr: *mut u32,
    sio_gpio_oe_set: *mut u32,
    timer_timerawl: *const u32,
}

impl Peripherals {
    fn new() -> Self {
        Self {
            resets_reset: (RESETS_BASE + RESETS_RESET_OFFSET) as *mut u32,
            resets_reset_done: (RESETS_BASE + RESETS_RESET_DONE_OFFSET) as *const u32,
            iobank0_io: (IO_BANK0_BASE + IO_BANK0_GPIO0_STATUS_OFFSET)
                as *mut [IoBank0HwIo; NUMBER_OF_GPIOS],
            padsbank0_io: (PADS_BANK0_BASE + PADS_BANK0_GPIO0_OFFSET)
                as *mut [u32; NUMBER_OF_GPIOS],
            sio_gpio_out_set: (SIO_BASE + SIO_GPIO_OUT_SET_OFFSET) as *mut u32,
            sio_gpio_out_clr: (SIO_BASE + SIO_GPIO_OUT_CLR_OFFSET) as *mut u32,
            sio_gpio_oe_set: (SIO_BASE + SIO_GPIO_OE_SET_OFFSET) as *mut u32,
            timer_timerawl: (TIMER_BASE + TIMER_TIMERAWL_OFFSET) as *const u32,
        }
    }
}

fn start_subsystems(p: &Peripherals) {
    // Subsystems that will be used
    let mask = RESETS_RESET_TIMER_BITS | RESETS_RESET_PIO0_BITS | RESETS_RESET_PADS_BANK0_BITS | RESETS_RESET_IO_BANK0_BITS;

    unsafe {
        // Take subsystems out of reset state
        core::ptr::write_volatile(p.resets_reset, core::ptr::read_volatile(p.resets_reset) & !mask);

        // Wait until reset state has been left
        while (core::ptr::read_volatile(p.resets_reset_done) & mask) != mask {}
    };
}

fn gpio_init_as_sio_output(p: &Peripherals, gpio: usize) {
    // Set input enable on, output disable off
    unsafe {
        core::ptr::write_volatile(
            &mut (*p.padsbank0_io)[gpio],
            (core::ptr::read_volatile(&(*p.padsbank0_io)[gpio])
                & (PADS_BANK0_GPIO0_IE_BITS | PADS_BANK0_GPIO0_OD_BITS))
                | PADS_BANK0_GPIO0_IE_BITS,
        )
    };

    // Zero all fields apart from fsel; we want this IO to do what the peripheral tells it.
    // This doesn't affect e.g. pullup/pulldown, as these are in pad controls.
    unsafe {
        core::ptr::write_volatile(
            &mut (*p.iobank0_io)[gpio].ctrl,
            GPIO_FUNC_SIO << IO_BANK0_GPIO0_CTRL_FUNCSEL_LSB,
        )
    };

    // Setup GPIO as output
    unsafe { core::ptr::write_volatile(p.sio_gpio_out_clr, 1 << gpio) };
    unsafe { core::ptr::write_volatile(p.sio_gpio_oe_set, 1 << gpio) };
}

fn gpio_put(p: &Peripherals, gpio: usize, value: bool) {
    let mask = 1 << gpio;
    if value {
        unsafe { core::ptr::write_volatile(p.sio_gpio_out_set, mask) };
    } else {
        unsafe { core::ptr::write_volatile(p.sio_gpio_out_clr, mask) };
    }
}

fn sleep_ms(p: &Peripherals, delay_ms: u32) {
    let base = unsafe { core::ptr::read_volatile(p.timer_timerawl) };
    while (unsafe { core::ptr::read_volatile(p.timer_timerawl) } - base) < (delay_ms * 1000) {}
}

pub fn blink_loop() -> ! {
    let p = Peripherals::new();
    const LED_PIN: usize = PICO_DEFAULT_LED_PIN;

    start_subsystems(&p);
    gpio_init_as_sio_output(&p, LED_PIN);
    loop {
        gpio_put(&p, LED_PIN, true);
        sleep_ms(&p, 500);
        gpio_put(&p, LED_PIN, false);
        sleep_ms(&p, 500);
    }
}
